File: app/src/main/java/com/example/algorandcarsharing/MainActivity.java
package com.example.algorandcarsharing;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import com.example.algorandcarsharing.databinding.ActivityMainBinding;
import com.example.algorandcarsharing.models.AccountModel;
import com.example.algorandcarsharing.services.IndexerService;
import com.google.android.material.navigation.NavigationView;
import com.google.android.material.snackbar.Snackbar;

import org.bouncycastle.jce.provider.BouncyCastleProvider;

import java.security.Security;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MainActivity extends AppCompatActivity {

    private AppBarConfiguration mAppBarConfiguration;
    private ActivityMainBinding binding;

    private IndexerService indexerService;

    private NavController navController;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Security.removeProvider("BC");
        Security.insertProviderAt(new BouncyCastleProvider(), 0);

        ExecutorService mExecutor = Executors.newSingleThreadExecutor();
        indexerService = new IndexerService();

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setSupportActionBar(binding.appBarMain.toolbar);
        binding.appBarMain.fab.setOnClickListener(view -> {
            Intent intent = new Intent(this, TripCreate.class);
            startActivity(intent);
        });

        DrawerLayout drawer = binding.drawerLayout;
        NavigationView navigationView = binding.navView;
        // Passing each menu ID as a set of Ids because each
        // menu should be considered as top level destinations.
        mAppBarConfiguration = new AppBarConfiguration.Builder(
                R.id.nav_home, R.id.nav_trips_created, R.id.nav_trips_joined, R.id.nav_account)
                .setOpenableLayout(drawer)
                .build();
        navController = Navigation.findNavController(this, R.id.nav_host_fragment_content_main);
        NavigationUI.setupActionBarWithNavController(this, navController, mAppBarConfiguration);
        NavigationUI.setupWithNavController(navigationView, navController);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        checkAccount();
        if(item.getItemId() == R.id.action_settings) {
            Intent intent = new Intent(this, SettingsActivity.class);
            startActivity(intent);
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onSupportNavigateUp() {
        NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment_content_main);
        return NavigationUI.navigateUp(navController, mAppBarConfiguration) || super.onSupportNavigateUp();
    }

    private void checkAccount() {
        AccountModel account = new AccountModel();
        try {
            account.loadFromStorage(this);
        }
        catch (Exception e) {
            e.printStackTrace();
        }
        if(account.getAddress() == null) {
            navController.navigate(R.id.nav_account);
        }
    }
}

File: app/src/main/java/com/example/algocarsharing/MainActivity.kt
package com.example.algocarsharing

import android.os.Bundle
import android.util.Log
import android.view.Menu
import androidx.appcompat.app.AppCompatActivity
import androidx.drawerlayout.widget.DrawerLayout
import androidx.navigation.findNavController
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.navigateUp
import androidx.navigation.ui.setupActionBarWithNavController
import androidx.navigation.ui.setupWithNavController
import com.algorand.algosdk.v2.client.common.Client
import com.algorand.algosdk.v2.client.common.IndexerClient
import com.algorand.algosdk.v2.client.common.Response
import com.algorand.algosdk.v2.client.model.ApplicationsResponse
import com.algorand.algosdk.v2.client.model.Enums
import com.algorand.algosdk.v2.client.model.TransactionsResponse
import com.example.algocarsharing.databinding.ActivityMainBinding
import com.google.android.material.navigation.NavigationView
import com.google.android.material.snackbar.Snackbar
import kotlinx.coroutines.*
import kotlinx.coroutines.Dispatchers.IO
import org.json.JSONObject


class MainActivity : AppCompatActivity() {

    private lateinit var appBarConfiguration: AppBarConfiguration
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.appBarMain.toolbar)

        binding.appBarMain.fab.setOnClickListener { view ->
            try {
                CoroutineScope(IO).launch {
                    //val jsonResp = async(Dispatchers.IO) { searchTransactions() }.await()
                    val jsonResp = withContext(IO) { searchTransactions() }
                    Log.d("Indexer response", jsonResp.toString(2))
                }

                Snackbar.make(view, "Ok", Snackbar.LENGTH_LONG)
                    .setAction("Action", null).show()
            }
            catch (e: Exception) {
                Log.e("Indexer error", e.stackTraceToString())
            }
        }
        val drawerLayout: DrawerLayout = binding.drawerLayout
        val navView: NavigationView = binding.navView
        val navController = findNavController(R.id.nav_host_fragment_content_main)
        // Passing each menu ID as a set of Ids because each
        // menu should be considered as top level destinations.
        appBarConfiguration = AppBarConfiguration(setOf(
                R.id.nav_home, R.id.nav_gallery, R.id.nav_slideshow), drawerLayout)
        setupActionBarWithNavController(navController, appBarConfiguration)
        navView.setupWithNavController(navController)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onSupportNavigateUp(): Boolean {
        val navController = findNavController(R.id.nav_host_fragment_content_main)
        return navController.navigateUp(appBarConfiguration) || super.onSupportNavigateUp()
    }

    // utility function to connect to a node
    private fun connectToNetwork(): Client {
        val INDEXER_API_ADDR = "10.0.2.2"
        val INDEXER_API_PORT = 8980
        return IndexerClient(INDEXER_API_ADDR, INDEXER_API_PORT)
    }

    private suspend fun searchTransactions(): JSONObject {
        val note = "67c8df8c4a6ef03decdfd0f174d16641"
        val indexerClientInstance = connectToNetwork() as IndexerClient
        val response: Response<TransactionsResponse> = indexerClientInstance
            .searchForTransactions()
            .notePrefix(note.encodeToByteArray())
            .txType(Enums.TxType.APPL)
            .limit(100)
            .execute()

        if (!response.isSuccessful()) {
            Log.e("Indexer error: ", response.code().toString())
            throw Exception(response.message())
        }

        val jsonObj = JSONObject(response.body().toString())
        return jsonObj
    }

    private suspend fun searchApplication(id: Int): JSONObject {
        val indexerClientInstance = connectToNetwork() as IndexerClient
        val response: Response<ApplicationsResponse> = indexerClientInstance
            .searchForApplications()
            .applicationId(id.toLong())
            .limit(1)
            .execute()

        if (!response.isSuccessful()) {
            Log.e("Indexer error: ", response.code().toString())
            throw Exception(response.message())
        }

        val jsonObj = JSONObject(response.body().toString())
        return jsonObj
    }
}